[14.0][MIG] pos_payment_terminal

alexis-via opened the pull request:
This PR works and has been tested on real Ingenico card reader (for France, with Telium) connected via pywebdriver (master branch).

It is a full re-write of the module, in order to take advantage of the new payment interface JS framework of Odoo POS.

FR translation updated.

Commit:
Initial check-in of the module pos_payment_terminal.

Commit:
Add dependancy on currency_iso_numeric.

Commit:
pos_customer_display bugfix : new massage on deleting orderLine and fixing unit_price computation on adding orderLine

Commit:
Fix copyright headers
PEP8 stuff

Commit:
pos_payment_terminal : bigger button for start transaction with touchscreen

Commit:
pos_payment_terminal: add missing css file
pos_customer_display: FIX JS code and make it more robust

Commit:
Remove currency_iso_numeric (replaced by pycountry)
Adapt JS code of pos_payment_terminal
Some cleanup

Commit:
Add Spanish translation by Alejandro Amador (thanks for your contribution !)
Move CSS definition from pos_payment_terminal_view.xml to pos_payment_terminal.xml
Small cleanups

Commit:
[FIX] Vicious bug that is invisible alone, but becomes visible when used with the module pos_pay_invoice

Commit:
PEP8

yvaucher committed:
Add license key in __openerp__.py

gurneyalex committed:
Add OCA as author of OCA addons

In order to get visibility on https://www.odoo.com/apps the OCA board has
decided to add the OCA as author of all the addons maintained as part of the
association.

Commit:
pos_payment_terminal: Add POT file, FR translation and demo data

Commit:
Better XML header

Commit:
pos_customer_display: Add button "Display Total to Customer" and port to new API
pos_payment_terminal: code cleanup

alexis-via committed:
pos_payment_terminal: smiplify JS code

Move description from __openerp__.py to README.rst
Update demo data

alexis-via committed:
FIX my previous commit: sent currency to POSbox

alexis-via committed:
[FIX] bug #36 : take advantage of the recent update in the odoo codebase to properly get the 'name' of the currency

sbidoul committed:
[UPD] prefix versions with 8.0

pedrobaeza committed:
[MIG] Make modules uninstallable

yuntux committed:
begin of porting pos_payment_terminal to Odoo 9.0

yuntux committed:
overload render_paymentlines

yuntux committed:
FIX : multiple calls to action on 'start transaction' click event

yuntux committed:
Add static/src/xml/templates.xml

pedrobaeza committed:
[MIG] Rename manifest files

alexis-via committed:
FIX crash on demo data

alexis-via committed:
Port pos_payment_terminal to v10

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

andreparames committed:
pos_payment_terminal: receive transaction refs

Provide a mechanism to send the order UID to the payment terminal,
which can then pass the transaction reference generated by the
payment provider back to Odoo, and which is then added to the
order payment lines. This allows their subsequent reconciliation.
The order is also automatically validated when the payment
finishes.

Commit:
[FIX] fixes currencies with decimals <>2

oca-transbot committed:
OCA Transbot updated translations from Transifex

andreparames committed:
pos_payment_terminal: inform user of transaction status

andreparames committed:
pos_payment_terminal: use currency precision

andreparames committed:
pos_payment_terminal: auto-validate orders with mixed payments

sbidoul committed:
remove obsolete .pot files [ci skip]

oca-travis committed:
[UPD] Update pos_payment_terminal.pot

hparfr committed:
Fix JS loading : module should only be loaded in pos
Fix indentation

hparfr committed:
Increment version number

OCA-git-bot committed:
[ADD] icon.png

rousseldenis committed:
[FIX] pos_payment_terminal: Fix tests

oca-travis committed:
[UPD] Update pos_payment_terminal.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-10.0/pos-10.0-pos_payment_terminal
Translate-URL: https://translation.odoo-community.org/projects/pos-10-0/pos-10-0-pos_payment_terminal/

legalsylvain committed:
[MIG] pos_payment_terminal

legalsylvain committed:
[MIG][12.0] pos_payment_terminal

legalsylvain committed:
Update pos_payment_terminal/static/src/js/devices.js

Co-Authored-By: Iván Todorovich <ivan.todorovich@druidoo.io>

legalsylvain committed:
[REF] rename field

legalsylvain committed:
fixup! [REF] rename field

legalsylvain committed:
fixup! fixup! [REF] rename field

oca-travis committed:
[UPD] Update pos_payment_terminal.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_payment_terminal
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_payment_terminal/

oca-travis committed:
[UPD] Update pos_payment_terminal.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_payment_terminal
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_payment_terminal/

danimv5 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (10 of 10 strings)

Translation: pos-12.0/pos-12.0-pos_payment_terminal
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_payment_terminal/es/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_payment_terminal
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_payment_terminal/

legalsylvain committed:
[IMP] pos_payment_terminal : improve display of the loading button

oca-travis committed:
[UPD] Update pos_payment_terminal.pot

OCA-git-bot committed:
pos_payment_terminal 12.0.0.1.2

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_payment_terminal
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_payment_terminal/

legalsylvain committed:
[IMP] pos_payment_terminal : improve the display of the button start transaction

oca-travis committed:
[UPD] Update pos_payment_terminal.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_payment_terminal
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_payment_terminal/

oca-travis committed:
[UPD] Update pos_payment_terminal.pot

OCA-git-bot committed:
pos_payment_terminal 12.0.0.1.3

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_payment_terminal
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_payment_terminal/

oca-travis committed:
[UPD] Update pos_payment_terminal.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-12.0/pos-12.0-pos_payment_terminal
Translate-URL: https://translation.odoo-community.org/projects/pos-12-0/pos-12-0-pos_payment_terminal/

alexis-via committed:
[MIG] pos_payment_terminal to v14

It is a full re-write of the module, in order to take advantage of the new
payment interface JS framework of Odoo POS
Update FR translation

alexis-via committed:
pos_payment_terminal: black, isort and pre-commit stuff

alexis-via committed:
pos_payment_terminal: update README and screenshots

legalsylvain commented on pos_payment_terminal/static/src/js/payment_terminal.js:
```diff
@@ -20,56 +20,60 @@ odoo.define('pos_payment_terminal.payment', function (require) {
             var order = this.pos.get_order();
             var pay_line = order.selected_paymentline;
             var currency = this.pos.currency;
             if (pay_line.amount < 0) {
```
maybe ``<=`` here, don't you think ?

sbidoul replied:
Done in https://github.com/akretion/pos/pull/20

legalsylvain commented on pos_payment_terminal/views/assets.xml:
```diff
@@ -3,10 +3,10 @@
             <script type="text/javascript" src="/pos_payment_terminal/static/src/js/models.js"></script>
-            <script type="text/javascript" src="/pos_payment_terminal/static/src/js/devices.js"></script>
-            <link rel="stylesheet" href="/pos_payment_terminal/static/src/css/pos_payment_terminal.css" />
+                    <!--            <script type="text/javascript" src="/pos_payment_terminal/static/src/js/devices.js"></script>
```
should be removed, I guess.

legalsylvain reviewed (commented):
Huge refactor ! Thanks a lot to begin to port PoS module in the new framework. 
I did a superficial review and LGTM. 
As far as I understand, we don't need now specific button to send to the terminal ? (I see you removed all the xml file).
Side question : do you like the new framework ? is it better than the previous one ?

thanks.

ivantodorovich commented on pos_payment_terminal/models/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from . import pos_payment_method
+
+# from . import pos_order
```
Is it required or deprecated?
if it's deprecated we should remove the file entirely

alexis-via replied:
I don't know. I didn't write that code; it was written by Acsone. I don't what is the purpose and how it works. That's why I just commented it and I didn't want to remove it.

sbidoul replied:
IIRC this was done to save the payment transaction id in the database when the terminal driver returns it. It [looks like there is native support for that in the point_of_sale](https://github.com/odoo/odoo/blob/66b468626e046a13a463cff3a2285801016f3ed7/addons/point_of_sale/static/src/js/payment.js#L48-L51) module so we should probably use that.

sbidoul replied:
I removed this code in https://github.com/akretion/pos/pull/20

ivantodorovich commented on pos_payment_terminal/tests/__init__.py:
```diff
@@ -0,0 +1 @@
+# from . import test_transactions
```
to check

sbidoul replied:
also removed in https://github.com/akretion/pos/pull/20

alexis-via committed:
pos_payment_terminal: remove commented lines

sbidoul commented:
@alexis-via this is not finished, right ?

This module needs to support [send_payment_cancel](https://github.com/odoo/odoo/blob/66b468626e046a13a463cff3a2285801016f3ed7/addons/point_of_sale/static/src/js/payment.js#L72) and [send_payment_request must return a promise](https://github.com/odoo/odoo/blob/66b468626e046a13a463cff3a2285801016f3ed7/addons/point_of_sale/static/src/js/payment.js#L54) that manages the transaction status via the `change:status` event and the [payment transactions list](https://github.com/akretion/pywebdriver/blob/a16d833fbcf7bbaa6370b1f61f66c7e0f38a45c9/pywebdriver/plugins/payment_base_driver.py#L59) we get in it.

Are you or @PierrickBrun working on this these days ? Shall we coordinate or efforts ?

sbidoul committed:
pos_payment_terminal: fix lint issues

sbidoul committed:
pos_payment_terminal: api conformity

Have send_payment_request return a promise as requested by the
documentation. This open the door to handling the transaction status
reported by the terminal driver asynchronously.

Also, catch errors when sending the transaction start command.

sbidoul commented:
@alexis-via I made this PR green in https://github.com/akretion/pos/pull/20 and improved it a little bit.

Now I'll work on handling the transaction statuses from the payment terminal drivers, compatible with [pywebdriver](https://github.com/akretion/pywebdriver/blob/master/pywebdriver/plugins/payment_mock_driver.py).

sbidoul committed:
pos_payment_terminal: reject transactions with zero amount

sbidoul committed:
pos_payment_terminal: remove dead code

Commit:
Merge pull request #20 from acsone/14.0-mig-pos-payment-terminal-sbi

[14.0] pos_payment_terminal migration improvements

sbidoul commented:
@ivantodorovich @legalsylvain your review comments have been processed.

sbidoul commented:
BTW, when the transaction has been sent to the terminal, this is what happens in the POS - I post this so others don't have to research why this is normal :)

![image](https://user-images.githubusercontent.com/692075/108195278-d6640e00-7117-11eb-80f3-cce9b79b355e.png)

This is normal behaviour until we have transaction status from the driver: the user has to manually confirm that the transaction suceeded by clicking "Force done". The "Connection error" message is hard coded in Odoo.

ivantodorovich reviewed (approved):
LGTM! :+1: 

Please squash migration commits before merging

sbidoul commented:
> Please squash migration commits before merging

Well, there are two authors and this is more a rewrite than a migration, and the commit history helps understand why things are as they are.

ivantodorovich commented:
> > Please squash migration commits before merging
> 
> Well, there are two authors and this is more a rewrite than a migration, and the commit history helps understand why things are as they are.

I'd at least clean a bit the commits, whilst keeping the authors

For instance there are commits like
- `pos_payment_terminal: remove commented lines`,
- `pos_payment_terminal: remove dead code` , 
- `pos_payment_terminal: fix lint issues`, 
- `pos_payment_terminal: black, isort and pre-commit stuff`



----

It's also a good opportunity to clean the previous history.
There are things like this that could be squashed:

![image](https://user-images.githubusercontent.com/1914185/108196389-d18f6200-70f7-11eb-80dc-c9f74d0f3350.png)

as well as the usual administrative commits

> Squash administrative commits (if any) with the previous commit for reducing commit noise. They are named as "[UPD] README.rst", "[UPD] Update $MODULE.pot", "Update translation files" and similar names, and comes from OCA-git-bot, oca-travis or oca-transbot. IMPORTANT: Don't squash legit translation commits, authored by their translators, with the message "Translated using Weblate (...)".

https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests

sbidoul commented:
I try to make small commits that are easy to review, and also sufficiently atomic that they are easy to revert. I also strive to not rewrite past history. Anyway,  I can't do the squashing myself here as this is not a branch I own.

ivantodorovich commented:
I'm not blocking in this 👍🏻

alexis-via commented:
/ocabot merge nobump
